src/sgame/components/SpikerComponent.cpp

```diff
@@ -87,44 +65,75 @@ void SpikerComponent::Think(int timeDelta) {
-		float relativeDamage = expectedDamage / durability;
-
-		if (onSameTeam) {
-			friendlyDamage += relativeDamage;
-		} else {
-			if (distance < SPIKER_SENSE_RANGE) {
-				sensing = true;
-				if (sensing && !lastSensing) {
-					RegisterFastThinker();
-				}
+		float damage = (targetArea / effectArea) * (float)SPIKER_MISSILES *
+			(float)BG_Missile(MIS_SPIKER)->damage;
+
+		expectedDamage += damage;
+
+		if (distance < SPIKER_SENSE_RANGE) {
+			sensing = true;
+
+			if (!lastSensing) {
+				logger.Verbose("Spiker #%i now senses an enemy and will check more frequently for "
+					"the best moment to shoot.", entity.oldEnt->s.number);
+
+				RegisterFastThinker();
 			}
-			enemyDamage += relativeDamage;
 		}
 	});
 
-	// Friendly entities that can receive damage substract from the scoring, enemies add to it.
-	float scoring = enemyDamage - friendlyDamage;
+	// Shoot if a viable target leaves sense range even if no damage is expected anymore. This
+	// guarantees that the spiker always shoots eventually when an enemy gets close enough to it.
+	bool senseLost = lastSensing && !sensing;
+
+	if (sensing || senseLost) {
+		bool lessDamage = (expectedDamage <= lastExpectedDamage);
 
-	// Shoot if a viable target leaves sense range even if the new scoring is bad.
-	// Guarantees that the spiker always shoots eventually when an enemy gets close enough to it.
-	bool senseLost = lastScoring > 0.0f && lastSensing && !sensing;
+		if (sensing) {
+			logger.Verbose("%i: Spiker #%i senses an enemy and expects to do %.1f damage.%s",
+				level.time, entity.oldEnt->s.number, expectedDamage, lessDamage
+				? " This has not increased, so it's time to shoot." : "");
+		}
 
-	if (scoring > 0.0f || senseLost) {
-		logger.Verbose("Spiker #%i scoring %.3f - %.3f = %.3f%s%s",
-		               entity.oldEnt->s.number, enemyDamage, friendlyDamage, scoring,
-		               sensing ? " (sensing)" : "", senseLost ? " (lost target)" : "");
+		if (senseLost) {
+			logger.Verbose("Spiker #%i lost track of all enemies after expecting to do %.1f damage."
+				" This makes the spiker angry, so it will shoot anyway.",
+				entity.oldEnt->s.number, lastExpectedDamage);
+		}
 
-		if ((scoring <= lastScoring && sensing) || senseLost) {
+		if ((sensing && lessDamage) || senseLost) {
 			Fire();
 		}
 	}
 
-	lastScoring = scoring;
+	lastExpectedDamage = expectedDamage;
 	lastSensing = sensing;
 }
 
+// TODO: Use a proper trajectory trace, once available, to ensure friendly buildables are never hit.
+bool SpikerComponent::SafeToShoot(Vec3 direction) {
+	const missileAttributes_t* ma = BG_Missile(MIS_SPIKER);
+	trace_t trace;
+	vec3_t mins, maxs;
+	Vec3 end = Vec3::Load(entity.oldEnt->s.origin) + (SPIKER_SPIKE_RANGE * direction);
+
+	// Test once with normal and once with inflated missile bounding box.
+	for (int i = 0; i <= 1; i++) {
+		float traceSize = (float)ma->size * ((1.0f-(float)i) + (float)i*SAFETY_TRACE_FUDGE_FACTOR);
```

Comment:
How about `for (float traceSize : {float(ma->size), ma->size * SAFETY_TRACE_FUDGE_FACTOR})`

Reply:
Thanks, will do!t